Clean code modifications merged

Tiko7454 opened the pull request.

Tiko7454 committed:
moved the whole logic to a singleton class

Tiko7454 committed:
class was seperated by two files

Tiko7454 committed:
linking the sentence_word_counter library

Tiko7454 committed:
project architecture changes

Tiko7454 committed:
cmake changes

Tiko7454 committed:
Added facilities and removed inconvenient tools
is_in was replaced with is_operator
added abstract class Convertor
global OPERATORS was moved to Convertor's protected field
                 and renamed to _OPERATORS
added is_special_symbol to detect operators and parentheses

Tiko7454 committed:
minor changed caused by file renaming

Tiko7454 committed:
added a class to cover infix->postfix convertion

Tiko7454 committed:
added a class to cover postfix->infix convertion

Tiko7454 committed:
main is modified to use polymorphism

Aurikon commented on postfix2infix/convertor.cxx:
```diff
@@ -0,0 +1,20 @@
+const std::array<char, 4> Convertor::_OPERATORS{'+', '-', '*', '/'};
+
+bool Convertor::is_operator(char ch) {
+    for (char el : _OPERATORS) {
```
use std::any_of

Aurikon commented on postfix2infix/convertor.cxx:
```diff
@@ -0,0 +1,20 @@
+    return false;
+}
+
+bool Convertor::is_operator(std::string ch) {
```
const reference

Aurikon commented on postfix2infix/infix_to_postfix_convertor.cxx:
```diff
@@ -0,0 +1,98 @@
+            continue;
+        }
+        if (not sub_expr.empty()) {
+            // this code is never reached
```
is this comment actual?

Aurikon replied:
is this comment actual?

Aurikon commented on postfix2infix/infix_to_postfix_convertor.cxx:
```diff
@@ -0,0 +1,98 @@
+
+void InfixToPostfixConvertor::_write_operand(std::string &token) {
+    _operands.push(token);
+    token = "";
```
token.clear()

Aurikon commented on postfix2infix/infix_to_postfix_convertor.cxx:
```diff
@@ -0,0 +1,98 @@
+void InfixToPostfixConvertor::_write_sub_expr(std::string &sub_expr) {
+    InfixToPostfixConvertor convertor{sub_expr};
+    _operands.push(convertor.convert());
+    sub_expr = "";
```
sub_expr.clear()

Tiko7454 committed:
code optimised
_OPERATORS were sorted to perform binary_search
is_operator's argument is passed by constant referance

Tiko7454 committed:
code optimisation
moved some part of code to the beginning of the loop (to increase performance)
removed depricated code

Tiko7454 committed:
code optimisations
